shared/trust/hmac.go

```diff
@@ -0,0 +1,205 @@
+package trust
+
+import (
+	"bytes"
+	"crypto/hmac"
+	"crypto/sha256"
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"hash"
+	"io"
+	"net/http"
+	"strings"
+
+	"github.com/canonical/lxd/shared/api"
+)
+
+// HMACFormatter represents arbitrary formats to diplay and parse the actual HMAC.
+// For example implementations like argon2 extend the format with an additional salt.
+// Example using argon2: `Authorization: <version> <salt>:<HMAC>`.
+type HMACFormatter interface {
+	// The Write* methods allow the creation of an HMAC based on various inputs.
+	WriteBytes(b []byte) ([]byte, error)
+	WriteJSON(v any) ([]byte, error)
+	WriteRequest(r *http.Request) ([]byte, error)
+
+	splitVersionFromHMAC(header string) (HMACVersion, string, error)
```

Comment:
Does this need to be in the interface? Its not exported.

Reply:
No this doesn't need to be in the interface.
I put it there as I wanted to access it from all the implementations of the interface but realized it's enough to have it as part of the base struct as this one is already embedded in the argon struct.

Fixed and removed.